server/data/data-jmap/src/test/java/org/apache/james/jmap/api/change/MailboxChangeRepositoryContract.java

```diff
@@ -0,0 +1,124 @@
+
+        MailboxChange change = MailboxChange.createdChange(null, 1L, UUID.randomUUID());
+
+        assertThatThrownBy(() -> repository.save(change)).isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    default void saveChangeShouldFailWhenNoState() {
+        MailboxChangeRepository repository = mailboxChangeRepository();
+
+        MailboxChange change = MailboxChange.createdChange(null, 1L, UUID.randomUUID());
+
+        assertThatThrownBy(() -> repository.save(change)).isInstanceOf(NullPointerException.class);
+    }
+
+    @Test
+    default void getChangesShouldSuccess() {
+        MailboxChangeRepository repository = mailboxChangeRepository();
+
+        MailboxChange change = MailboxChange.createdChange(ACCOUNT_ID, 1L, UUID.randomUUID());
+        repository.save(change);
+
+        assertThat(repository.getFromState(ACCOUNT_ID, 0L, Optional.empty())).containsExactly(change);
+    }
+
+    @Test
+    default void getChangesShouldReturnEmptyWhenNoNewerState() {
+        MailboxChangeRepository repository = mailboxChangeRepository();
+
+        MailboxChange change = MailboxChange.createdChange(ACCOUNT_ID, 1L, UUID.randomUUID());
+        repository.save(change);
+
+        assertThat(repository.getFromState(ACCOUNT_ID, 2L, Optional.empty())).isEmpty();
+    }
+
+    @Test
+    default void getChangesShouldPrioritizeNewerState() {
+        MailboxChangeRepository repository = mailboxChangeRepository();
+
+        MailboxChange change1 = MailboxChange.createdChange(ACCOUNT_ID, 1L, UUID.randomUUID());
+        MailboxChange change2 = MailboxChange.updatedChange(ACCOUNT_ID, 2L, UUID.randomUUID());
+        repository.save(change1);
+        repository.save(change2);
+
+        assertThat(repository.getFromState(ACCOUNT_ID, 0L, Optional.empty())).containsExactly(change2, change1);
+    }
+
+    @Test
+    default void getChangesShouldLimitChanges() {
+        MailboxChangeRepository repository = mailboxChangeRepository();
+
+        MailboxChange change1 = MailboxChange.createdChange(ACCOUNT_ID, 1L, UUID.randomUUID());
+        MailboxChange change2 = MailboxChange.updatedChange(ACCOUNT_ID, 2L, UUID.randomUUID());
+        MailboxChange change3 = MailboxChange.destroyedChange(ACCOUNT_ID, 3L, UUID.randomUUID());
+        repository.save(change1);
+        repository.save(change2);
+        repository.save(change3);
+
+        assertThat(repository.getFromState(ACCOUNT_ID, 0L, Optional.of(2))).containsExactly(change3, change2);
+    }
+
+    @Test
+    default void getChangesShouldLimitChangesWhenMaxChangesOmitted() {
+        MailboxChangeRepository repository = mailboxChangeRepository();
+        LongStream.range(0, 20).forEach(value -> {
+            MailboxChange change = MailboxChange.createdChange(ACCOUNT_ID, value, UUID.randomUUID());
+            repository.save(change);
+        });
+
+        assertThat(repository.getFromState(ACCOUNT_ID, 0L, Optional.empty())).hasSize(DEFAULT_NUMBER_OF_CHANGES);
+    }
```

Comment:
You forgot

```
        repository.save(change1);
        repository.save(change2);
        repository.save(change3);

        assertThat(repository.getFromState(ACCOUNT_ID, 1L, Optional.empty())).containsExactly(change2, change3);
```

Reply:
You mean assertion on the content?

Reply:
That when I say "from state2" state1 is ommitted.